src/JanusStream.cpp

```diff
@@ -0,0 +1,122 @@
+/**
+ * @file JanusStream.h
+ * @author Daniel Stiner (danstiner@gmail.com)
+ * @date 2021-03-10
+ * @copyright Copyright (c) 2021 Daniel Stiner
+ */
+
+#include "JanusStream.h"
+
+#pragma region Constructor/Destructor
+JanusStream::JanusStream(
+    ftl_channel_id_t channelId,
+    ftl_stream_id_t streamId,
+    MediaMetadata mediaMetadata) :
+    channelId(channelId),
+    streamId(streamId),
+    mediaMetadata(mediaMetadata)
+{ }
+#pragma endregion
+
+#pragma region Public methods
+void JanusStream::SendRtpPacket(const std::vector<std::byte>& packet)
+{
+    std::lock_guard lock(mutex);
+
+    for (const auto& session : viewerSessions)
+    {
+        session->SendRtpPacket(packet, mediaMetadata);
+    }
+
+    for (const auto& relay : relays)
+    {
+        relay.Client->RelayPacket(packet);
+    }
+}
+
+void JanusStream::AddViewerSession(JanusSession* session)
+{
+    std::lock_guard lock(mutex);
+    viewerSessions.insert(session);
+}
+
+size_t JanusStream::RemoveViewerSession(JanusSession* session)
+{
+    std::lock_guard lock(mutex);
+    return viewerSessions.erase(session);
+}
+
+std::unordered_set<JanusSession*> JanusStream::RemoveAllViewerSessions()
+{
+    std::lock_guard lock(mutex);
+    std::unordered_set<JanusSession*> viewerSessions;
+    viewerSessions.swap(viewerSessions);
+    return viewerSessions;
+}
+
+size_t JanusStream::GetViewerCount() const
+{
+    std::lock_guard lock(mutex);
+    return viewerSessions.size();
+}
+
+void JanusStream::AddRelayClient(const std::string targetHostname,
+    std::unique_ptr<FtlClient> client)
+{
+    std::lock_guard lock(mutex);
+    relays.push_back(Relay { .TargetHostname = targetHostname, .Client = std::move(client) });
+}
+
+size_t JanusStream::StopRelay(const std::string& targetHostname)
+{
+    std::list<Relay> removedRelays;
+    {
+        std::lock_guard lock(mutex);
+        for (auto it = relays.begin();
+            it != relays.end();)
+        {
+            if (it->TargetHostname == targetHostname) {
+                removedRelays.push_back(std::move(*it));
+                it = relays.erase(it);
+            }
+        }
+    }
+    for (Relay& relay : removedRelays)
+    {
+        spdlog::info("Stopping relay for channel {} / stream {} -> {}...",
+            channelId, streamId, relay.TargetHostname);
+         relay.Client->Stop();
+    }
+    return removedRelays.size();
+}
+
+void JanusStream::StopRelays()
+{
+    std::lock_guard lock(mutex);
```

Comment:
This is my only remaining concern, and you said something over voip about this being reasonable, so it could be that my concern is not warranted. :)

Reply:
(concern being holding a lock while stopping every individual relay)

Reply:
I think my justification was if this is only called when we're killing the stream, so we don't care about delaying packets. But might as well clean it up.

Reply:
Cleaned it up with some swaps